scheduler/scheduler-server/src/test/java/functionaltests/recover/TaskReconnectionToDownNodeTest.java

```diff
@@ -0,0 +1,247 @@
+        JobResult jobResult = scheduler.getJobResult(jobid);
+        Assert.assertFalse(jobResult.hadException());
+    }
+
+    private void waitForJobToFinish(JobId jobid) throws Exception {
+        schedulerHelper.waitForEventJobFinished(jobid);
+    }
+
+    private void checkReplicateTasks(JobState jobState) {
+        int maximumNumberOfPendingReplicateTasks = jobState.getNumberOfPendingTasks() - 1;
+
+        for (int i = 1; i <= maximumNumberOfPendingReplicateTasks; i++) {
+
+            TaskState taskState = jobState.getTasks().get(i);
+            TaskStatus taskStatus = taskState.getTaskInfo().getStatus();
+
+            assertThat(taskStatus.equals(TaskStatus.PENDING) || taskStatus.equals(TaskStatus.RUNNING)).isTrue();
+
+            if (taskStatus.equals(TaskStatus.RUNNING)) {
+                this.checkThatExecutionHostnameIsDifferent(taskState);
+            }
+        }
+    }
+
+    private void checkThatExecutionHostnameIsDifferent(TaskState taskState) {
+        String formerExecutionHost = this.taskExecutionHostnamePerTaskId.get(taskState.getTaskInfo()
+                                                                                      .getTaskId()
+                                                                                      .longValue());
+        String currentExecutionHost = taskState.getTaskInfo().getExecutionHostName();
+
+        assertThat(formerExecutionHost).isNotEqualTo(currentExecutionHost);
+    }
+
+    private void checkFirstAndLastTask(JobState jobState) {
+        // we have exactly one task that is finished (the first one)
+        assertThat(jobState.getNumberOfFinishedTasks()).isEqualTo(1);
+        // at least we have the last task that is pending (the merge task)
+        assertThat(jobState.getNumberOfPendingTasks()).isAtLeast(1);
+    }
+
+    private void recordTaskExecutionHostInfo(JobId jobid) throws Exception {
+        for (int i = 0; i < NUMBER_OF_REPLICATE_TASKS; i++) {
+            TaskState taskState = schedulerHelper.getSchedulerInterface().getJobState(jobid).getTasks().get(i);
+            TaskInfo taskInfo = taskState.getTaskInfo();
+            this.taskExecutionHostnamePerTaskId.put(taskInfo.getTaskId().longValue(), taskInfo.getExecutionHostName());
+        }
+    }
+
+    private void restartScheduler() throws Exception {
+        schedulerHelper = new SchedulerTHelper(false,
+                                               new File(this.getSchedulerReStartConfigurationURL()
+                                                            .toURI()).getAbsolutePath(),
+                                               new File(RM_CONFIGURATION_RESTART.toURI()).getAbsolutePath(),
+                                               null,
+                                               false);
+        this.createNodes();
+    }
+
+    private void killSchedulerAndNodes() throws Exception {
+        Thread.sleep(RESTART_SCHEDULER_INTER_TIME_IN_MILLISECONDS);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2925)